compilers/concrete-compiler/compiler/include/concretelang/Dialect/TFHE/IR/TFHETypes.td

```diff
@@ -28,7 +28,7 @@ def TFHE_GLWECipherTextType
   let parameters = (ins
                     // The size of the mask
                     "mlir::concretelang::TFHE::GLWESecretKey"
-                    : $key);
+                    : $key, DefaultValuedParameter<"int", "-1">: $bitWidth);
```

Comment:
I've talked with Alex as I created this PR, so he's aware. The reason, I think, was that it wasn't useful. Not it is.

Reply:
I'm not sure it's a good idea because this bit width information is required for pretty much all primitive ops, and it's not just the result, it's also the intermediates. It's much more tedious to implement (need to change every single `builder.create<TFHE::SomeOp>`), comapred to just changing `GLWECipherTextType::get(...)`, which this commit does with `Showing 8 changed files with 50 additions and 20 deletions.`.

Reply:
Also, it's safer as we can't forget it :slightly_smiling_face:

Reply:
but we shouldn't use the dialect as a way to propagate an information that can be done otherwise.
> bit width information is required for pretty much all primitive ops

isn't it just for bootstrap?

Reply:
> isn't it just for bootstrap?

Nope, we treat every primitive operation the same.

> but we shouldn't use the dialect as a way to propagate an information that can be done otherwise.

The observations remain, if we decide to go with the attribute, 100s of lines of code would need to be changed, and need to be maintained with care (e.g., it's super easy to introduce a conversion but forget to set the attribute). If we do in the type, it would take less work and it'll be more resilient to developer memory.

Also, I don't think it has any drawbacks. It doesn't change parsing, and if you want to create the type using code, you **need to** think about what's the appropriate thing and if there isn't you can use -1.

Reply:
> Nope, we treat every primitive operation the same.

Does it mean that we will also have a `additionCountPerBitWidth` in the statistics ? 

> 100s of lines of code would need to be changed

That's overestimated, searching for the string `create<TFHE::` in the codebase shows 26 occurences, while `GLWECipherTextType::get` shows 12 occurences. It's the same order of magnitude in locs I would say.

> (e.g., it's super easy to introduce a conversion but forget to set the attribute)

Yes, but as soon as your `.getAttr("input_bitwidth")` will fail in the statistics computation, the developer will know that he missed something. Given the size of the test suite, I think it is fair to say that we __will__ catch such a scenario. 

> If we do in the type, it would take less work and it'll be more resilient to developer memory.

Other developers won't have the memory of why you brought that, and sooner than later, everybody will have lost sight of why this was brought back in, and what relies on it. It's going to be very likely for a developer that does not have the knowledge you have, to make mistakes when using that. It happened to me quite a few time already since I started to work on the compiler, and took time to debug these things. 

Put differently, not adding this will make things more straightforward. Since we are many people to work on overlapping sections of the compiler, I would say that this is worth the effort.

Reply:
> Does it mean that we will also have a additionCountPerBitWidth in the statistics ?

Yes, it'll be in the PR, I have it locally.

> searching for the string create<TFHE:: in the codebase shows 26 occurences

That's only the ones using the namespace, if you search for `mlir::concretelang::TFHE` you'll see A LOT more (and hidden ones such as the ones using just the type name, and we can't test for every possible conversion).

> while GLWECipherTextType::get shows 12 occurences

That is indeed only 12 occurences, which are updated using 70 lines of change. 

> as soon as your .getAttr("input_bitwidth") will fail in the statistics computation, the developer will know that he missed something. Given the size of the test suite, I think it is fair to say that we will catch such a scenario.

Yes but then they need to learn what that attribute is and properly set it, which is another source of confusion, compared to just using the GLWE type.

> Other developers won't have the memory of why you brought that

I'm putting a giant comment:
```
// bitWidth is optional and is used for extracting statistics
// see compiler/lib/Dialect/TFHE/Analysis/ExtractStatistics.cpp
```
I don't think anyone would forget it :slightly_smiling_face: 

> not adding this will make it more straightforward for people to build on top of this work

What does it make harder exactly? In the API, the only change is you need to call:
```
GLWECipherText::get(key, bitWidth)
```
instead of
```
GLWECipherText::get(key)
```
and it's much better than forgetting it because it's required for proper analysis. And if developers aren't sure why it's required, they can look at the definition and see the comment above.

Reply:
Well an attribute named `statistics_input_bitwidth` is pretty self-explanatory, and not nearly as invasive. 

Anyway, I think I made my point clear enough and I'm in holidays tonight so up to you guys.

Reply:
We've decided to continue with `per_parameter` statistics for this PR.